lib/tooling/llvm-mca-tool.ts

```diff
@@ -47,21 +47,64 @@ export class LLVMMcaTool extends BaseTool {
 
     override async runTool(compilationInfo: Record<any, any>, inputFilepath?: string, args?: string[]) {
         const isets = new InstructionSets();
-        const target = isets.getInstructionSetTarget(compilationInfo.compiler.instructionSet);
-        if (target != null) args?.push('-mtriple=' + target);
+        let target = isets.getInstructionSetTarget(compilationInfo.compiler.instructionSet);
+        const prependArgs: string[] = [];
 
+        // Check if compiler target is overridden with --target.
         for (const arg of compilationInfo.options) {
-            if (arg.startsWith('-mcpu') || arg.startsWith('-march') || arg.startsWith('-mtriple')) {
-                args?.push(arg);
+            if (arg.startsWith('--target=')) {
+                target = arg.replace(/^--target=/, '');
+            }
+        }
+
+        if (target != null) prependArgs.push('-mtriple=' + target);
+
+        let haveCPU = false;
+        let haveM32 = false;
+        for (const arg of compilationInfo.options) {
+            if (arg.startsWith('-march') && target?.startsWith('x86_64')) {
+                // On x86 in clang the CPU is specified with -march,
+                // but MCA requires -mcpu to function, so rewrite it.
+                // Note that -march in MCA means something quite
+                // different (it's a target) than it means with clang.
+                if (arg.startsWith('-march=x86-64')) {
+                    // This is an architecture level, not a CPU, so don't translate.
+                    continue;
+                }
+
+                prependArgs.push(arg.replace(/^-march=/, '-mcpu='));
+                haveCPU = true;
+            }
+            if (arg === '-m32') haveM32 = true;
+            if (arg.startsWith('-mcpu')) {
+                prependArgs.push(arg);
+                haveCPU = true;
             }
         }
 
         if (compilationInfo.filters.binary || compilationInfo.filters.binaryObject) {
             return this.createErrorResponse('<cannot run analysis on binary>');
         }
 
+        // If no other CPU specified, default to mcpu=generic to
+        // override llvm-mca's default of -mcpu=native, otherwise MCA's
+        // CPU analysis varies according to the machine MCA happens to
+        // run on which can change over time.
+        if (!haveCPU) {
+            prependArgs.push('-mcpu=generic');
+        }
+
+        if (target?.startsWith('x86_64') && haveM32) {
```

Comment:
So this handles one case where 32-bit mode is enabled via `-m32`. Need to check for another case when target triple starts with `i386`.

Reply:
Right, we don't need to handle `i386` triple case - we have a triple after all, passing it is enough.

Reply:
Btw, I'm using one-liner like this to test parameters `echo 'int main(int argc) { return argc * argc; }' | clang -x c - -target i386-pc-linux-gnu -S -o - | llvm-mca -mtriple=i386-pc-linux-gnu`